Export tags through ?tag all --json

OoLunar opened the pull request:
Same as `?tag all --text`, except it exports the content as JSON, with the tag's content. Closes #83. Example export:
```json
{"id":3,"name":"codeblocks","owner_id":336733686529654798,"uses":1,"can_delete":true,"is_alias":false,"content":"Use codeblocks for formatting code:\n\\`\\`\\`<language name>\n<your code here>\n\\`\\`\\`\n\nFor example:\n\\`\\`\\`python\nif True:\n    print(\"Hi!\")\n\\`\\`\\`\nProduces:\n```python\nif True:\n    print(\"Hi!\")\n```"}
```

OoLunar committed:
Added `?tag all --json` option

OoLunar committed:
Merge pull request #1 from OoLunar/export-tags-json

Added `?tag all --json` option

JeremyEspresso commented on cogs/tags.py:
```diff
@@ -870,6 +871,33 @@ async def _tag_all_text_mode(self, ctx):
+
+        for r in rows:
+            fp = io.BytesIO(r[0].encode())
+        await ctx.send(file=discord.File(fp, 'tags.txt'))
```
Shouldn't the extension be .json?

OoLunar replied:
Absolutely. I thought I did it, must've Ctrl + Z'd it

OoLunar committed:
Update tags.py

Changed `?tag all --json` to upload `tags.json` instead of `tags.txt` file.

OoLunar commented:
~~Just realized that `?tag all --json` only exports the latest tag instead of all of them. Fixing that now.~~

All tags are exported. Aliases are exported with the tags content.

OoLunar committed:
Update tags.py

All tags are now exported instead of one. Tags are now exported into an array. Thanks @IAmTomahawkx and @NatsumiHB!

bsrecker commented:
I only see the latest tag being exported instead of all of the tags with this code change.

bsrecker reviewed (approved):
Downloaded and verified that changes work for tag export to json.

AbstractUmbra commented on cogs/tags.py:
```diff
@@ -841,6 +841,7 @@ async def tags(self, ctx, *, member: TagMember = None):
     def _get_tag_all_arguments(args):
         parser = Arguments(add_help=False, allow_abbrev=False)
         parser.add_argument('--text', action='store_true')
+        parser.add_argument('--json', action='store_true')
```
This should perhaps be use [`ArgumentParser.add_mutually_exclusive_group`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) as currently you can use both flags and only the first will be accepted.

OoLunar replied:
Perhaps something like this? I don't typically program in Python, so I'm sure what I'm doing can be optimized. Thank you very much for linking the docs
```python
def _get_tag_all_arguments(args):
    parser = Arguments(add_help=False, allow_abbrev=False)
    parser.add_mutually_exclusive_group(required=False)
    parser.add_argument('--text', action='store_true')
    parser.add_argument('--json', action='store_true')
    if args is not None:
        return parser.parse_args(shlex.split(args))
    else:
        return parser.parse_args([])
 ```

AbstractUmbra replied:
`add_mutually_exclusive_group` returns the `group` that you should add the arguments to:

```py
    group = parser.add_mutually_exclusive_group(required=False)
    group.add_argument('--text', action='store_true')
    group.add_argument('--json', action='store_true')
```

OoLunar replied:
Just tested your suggestion. Everything seems good, I'll put it into the PR now

OoLunar committed:
Update tags.py

Resolves https://github.com/Rapptz/RoboDanny/pull/93#discussion_r589120172

OoLunar committed:
Merge branch 'Rapptz:rewrite' into rewrite

OoLunar committed:
Merge branch 'Rapptz:rewrite' into rewrite

OoLunar committed:
Merge branch 'Rapptz:rewrite' into rewrite

thecashewtrader commented:
Any update on this PR? This is a feature that would greatly help me.

OoLunar commented:
From what I understand, R. Danny is no longer being worked on and is only functioning until the message intent kills it. I don't think the PR will be merged

thecashewtrader commented:
That's sad tbh, R. Danny being phased out was one of the main reasons I wanted an export, so that I can have tags backed up in plaintext.

JDJGInc commented:
> From what I understand, R. Danny is no longer being worked on and is only functioning until the message intent kills it. I don't think the PR will be merged

R.danny has been verified, I think this pr is fine to be merged if Danny likes the concept and code.
